### DISCUSSION TITLE
Using Reaktoro for kinetic reactions in reactive transport simulations

### DISCUSSION THREAD

**Discussion by taj004** [reaktoro/reaktoro#321] [Q&A]
### Description

Hello. I am currently working on developing a reactive transport solver, using Reaktoro to solve chemical reaction equations. My question is, for problems with a mix of equilibrium and kinetics reactions, what is the correct approach in Reaktoro to solve the reaction equations, given transported elements (i.e. chemical element and charge)? To be specific, for only equilibrium reactions, I used the `reaktoro.EquilibriumConditions` to set the transported elements for the Reaktoro equilibrium solver (cf. https://reaktoro.org/tutorials/equilibrium/equilibrium-with-given-component-amounts.html). What is the equivalent approach for the kinetic solver? 
Below is an MWE to illustrate my question. Here, I attempt to solve both kinetic and equilibrium reactions (I make no assessment as to whether the setup is realistic). At each time step, I update the elements in the system. But as you see from the printing of `states`, the updated elements are not incorporated in the simulation.  

### Example

```Python
import reaktoro as rt
import numpy as np
import matplotlib.pyplot as plt

# For use random numbers later on
np.random.seed(1)

# Database
db = rt.SupcrtDatabase("supcrtbl")

# Parameters and reaction rates
params = rt.Params.embedded("PalandriKharaka.yaml")
reaction_rate_model = rt.ReactionRateModelPalandriKharaka(params)

# System
system = rt.ChemicalSystem(
    # The standard business
    db,
    rt.AqueousPhase().set( rt.ActivityModelPitzer() ),     
    rt.MineralPhases("Calcite Magnesite Dolomite"),
    
    # For mineral kinetic reactions
    rt.MineralReaction("Calcite").setRateModel(reaction_rate_model),
    rt.MineralReaction("Magnesite").setRateModel(reaction_rate_model),
    rt.MineralReaction("Dolomite").setRateModel(reaction_rate_model),    
    
    rt.MineralSurface("Calcite", 6.0, "m2/m3"),
    rt.MineralSurface("Magnesite", 6.0, "m2/m3"),
    rt.MineralSurface("Dolomite", 6.0, "m2/m3")
    )

#%% Prepare to solve

day = 1 * 60 * 60 * 24 # one day, in seconds
duration = 15 * day # final time
steps = 1000 # number of time steps
dt = duration / steps # time step size

table = rt.Table() # table to store simulation information

state = rt.ChemicalState(system)
state.set("H2O(aq)",   1.0, "kg")
state.set("Calcite",   1.0, "mol")
state.set("Magnesite", 1.0, "mol")
state.set("O2(aq)",    1.0, "umol") 

# Scale volumes, so the concentations are bulk concentations
state.scalePhaseVolume("AqueousPhase", 0.1, "m3")
state.scalePhaseVolume("Calcite", 0.45, "m3")
state.scalePhaseVolume("Magnesite", 0.45, "m3")

new_state = state.clone()

solver = rt.KineticsSolver(system)
cond = rt.EquilibriumConditions(system)

# Initial elements
b = new_state.componentAmounts().asarray()

#%% Solve

# Time loop
for i in range(steps + 1):
    
    # Let this represent `transported' elements (both kinetic and equilibrium)
    b += dt * 5e-7 * np.random.uniform(b.size)
   
    cond.setInitialComponentAmounts(b)
    
    # Solve the current system
    result = solver.solve(new_state, dt, cond) 
 
    assert result.succeeded(), f"calculation faild at time step {i}" 
    
    # Get current information about the system
    props = new_state.props() 
    aprops = rt.AqueousProps.compute(props)
    
    # Store in table
    table.column("Time") << i*dt / day # days
    
    table.column("Calcite") << props.speciesAmount("Calcite").val()
    table.column("Magnesite") << props.speciesAmount("Magnesite").val()
    table.column("Dolomite") << props.speciesAmount("Dolomite").val()
    table.column("Ca+2") << props.speciesAmount("Ca+2")
    table.column("Mg+2") << props.speciesAmount("Mg+2").val()
    
    table.column("rateCalcite") << props.reactionRate("Calcite")
    table.column("rateMagnesite") << props.reactionRate("Magnesite")
    table.column("rateDolomite") << props.reactionRate("Dolomite")
    
    table.column("ph") << aprops.pH() 
    
    table.column("H") << props.elementAmount("H").val()
    table.column("C") << props.elementAmount("C").val()
    
# end time-loop     

# Compare states
print("Initial state")
print(state)
print("Final state")
print(new_state)

#%% Plot

plt.figure()
plt.plot(table["Time"], table["Ca+2"], color="blue", label="$Ca^{2+}$")
plt.plot(table["Time"], table["Mg+2"], color="orange", label="$Mg^{2+}$")
plt.ylabel("Amount [mol]")
plt.xlabel("Time [day]")
plt.legend(loc="upper right")
plt.show()   

fig = plt.figure()
plt.plot(table["Time"], table["rateCalcite"], color="blue", label="Calcite")
plt.plot(table["Time"], table["rateDolomite"], color="orange", label="Dolomite")
plt.ylabel("Rate [mol/s]")
plt.xlabel("Time [day]")
plt.legend(loc="upper right")
plt.show()    
 
plt.Figure()
plt.plot(table["Time"], table["H"], label="H")
plt.plot(table["Time"], table["C"], label="C")
plt.legend(loc="upper right")
```
The print-lines produce
```Python
Initial state
+-------------------------+------------+------+
| Property                |      Value | Unit |
+-------------------------+------------+------+
| Temperature             |   298.1500 |    K |
| Pressure                |     1.0000 |  bar |
| Charge:                 | 3.9882e-14 |  mol |
| Element Amount:         |            |      |
| :: H                    | 1.1069e+04 |  mol |
| :: C                    | 2.8253e+04 |  mol |
| :: O                    | 9.0293e+04 |  mol |
| :: Mg                   | 1.6054e+04 |  mol |
| :: Ca                   | 1.2198e+04 |  mol |
| Species Amount:         |            |      |
| :: H2O(aq)              | 5.5345e+03 |  mol |
| :: CaOH+                | 9.9705e-15 |  mol |
| :: CO(aq)               | 9.9705e-15 |  mol |
| :: CO2(aq)              | 9.9705e-15 |  mol |
| :: CO3-2                | 9.9705e-15 |  mol |
| :: Ca(HCO3)+            | 9.9705e-15 |  mol |
| :: Ca+2                 | 9.9705e-15 |  mol |
| :: H+                   | 9.9705e-15 |  mol |
| :: H2(aq)               | 9.9705e-15 |  mol |
| :: HCO3-                | 9.9705e-15 |  mol |
| :: HO2-                 | 9.9705e-15 |  mol |
| :: Mg(HCO3)+            | 9.9705e-15 |  mol |
| :: Mg+2                 | 9.9705e-15 |  mol |
| :: MgOH+                | 9.9705e-15 |  mol |
| :: O2(aq)               | 9.9705e-05 |  mol |
| :: OH-                  | 9.9705e-15 |  mol |
| :: H2O2(aq)             | 9.9705e-15 |  mol |
| :: MgCO3(aq)            | 9.9705e-15 |  mol |
| :: CaCO3(aq)            | 9.9705e-15 |  mol |
| :: Calcite :: CaCO3     | 1.2198e+04 |  mol |
| :: Magnesite :: MgCO3   | 1.6054e+04 |  mol |
| :: Dolomite :: CaMgC2O6 | 1.0000e-16 |  mol |
+-------------------------+------------+------+
Final state
+-------------------------+-------------+------+
| Property                |       Value | Unit |
+-------------------------+-------------+------+
| Temperature             |    298.1500 |    K |
| Pressure                |      1.0000 |  bar |
| Charge:                 | -1.0235e-16 |  mol |
| Element Amount:         |             |      |
| :: H                    |  1.1069e+04 |  mol |
| :: C                    |  2.8253e+04 |  mol |
| :: O                    |  9.0293e+04 |  mol |
| :: Mg                   |  1.6054e+04 |  mol |
| :: Ca                   |  1.2198e+04 |  mol |
| Species Amount:         |             |      |
| :: H2O(aq)              |  5.5344e+03 |  mol |
| :: CaOH+                |  6.9441e-06 |  mol |
| :: CO(aq)               |  1.0000e-16 |  mol |
| :: CO2(aq)              |  2.2973e-06 |  mol |
| :: CO3-2                |  8.5703e-03 |  mol |
| :: Ca(HCO3)+            |  5.4894e-06 |  mol |
| :: Ca+2                 |  4.1667e-03 |  mol |
| :: H+                   |  8.2678e-09 |  mol |
| :: H2(aq)               |  1.0000e-16 |  mol |
| :: HCO3-                |  1.3344e-02 |  mol |
| :: HO2-                 |  1.0000e-16 |  mol |
| :: Mg(HCO3)+            |  2.2266e-05 |  mol |
| :: Mg+2                 |  1.7334e-02 |  mol |
| :: MgOH+                |  4.0889e-04 |  mol |
| :: O2(aq)               |  9.9705e-05 |  mol |
| :: OH-                  |  1.2960e-02 |  mol |
| :: H2O2(aq)             |  1.0000e-16 |  mol |
| :: MgCO3(aq)            |  1.1050e-03 |  mol |
| :: CaCO3(aq)            |  5.9200e-04 |  mol |
| :: Calcite :: CaCO3     |  1.2198e+04 |  mol |
| :: Magnesite :: MgCO3   |  1.6054e+04 |  mol |
| :: Dolomite :: CaMgC2O6 |  1.0262e-16 |  mol |
+-------------------------+-------------+------+
```

### What is your operating system?

Linux

**Comment by allanleal** [comment]
Hi @taj004, I'm on vacation today, but I can give you a quick reply. You can use `EquilibriumConditions` with `KineticsSolver` as you did above. The array `b` corresponds to the amounts of elements in the system, without distinguishing elements from species reacting with an instantaneous equilibrium model or with a kinetic rate model. Thus, for example, `b[Ca]` contains not only the amount of Ca in the aqueous phase, but also in the kinetically reacting minerals. The solver will respect these element amounts; the final reacted chemical state will conserve elemental mass (generally up to machine accuracy, ~1e-16). 

Thus, in your reactive transport model, you should keep track of `b` for each cell, which should be computed from contributions of fluid and solid species: `b = b(solid) + b(fluid)`. Your transport operation (without reactions) should produce new `b(fluid)` array for each grid point/cell while `b(solid)` array remains constant.

**Reply by allanleal** [reply]
I forgot to mention. When you have kinetic reactions in your chemical system, there are additional components beyond the elements, one for each kinetic reaction. Their values are the _extent of reaction_:

![image](https://github.com/reaktoro/reaktoro/assets/5825588/33ea3bbd-a5a2-4a09-b0b1-f3f939a08a58)

For your mineral reactions, the stoichiometric coefficient is -1 for the mineral and zero for all other species (as the minerals dissolve/precipitate, the aqueous species equilibrate). Thus, the extent of reaction, for these particular cases of reactions, is equal to the negative of the mineral amounts. I still need to write a paper detailing the kinetic algorithm to make these details simpler to understand. For now, it's important to note that `b = [b(elements), b(extent-of-reactions)]`, so in your transport model you should update the `b(elements)` entries above, leaving the last entries in `b` unchanged (as they are associated with non-flowing minerals).

**Reply by taj004** [reply]
Hello @allanleal , Thanks for your reply. 
So, if I have understood you correctly, my struggle with the `KineticSolver` is because my vector `b` only contains the _element_-part, and not the _extent of reaction_-part?

**Reply by taj004** [reply]
Also, what is the correct input to the `EquilibriumCondition`, and subsequently to the `KineticSolver`? I tested`b=[elements, -minerals]`, where minerals are the mineral amounts in the MWE, but the `EquililibriumCondition` returned the error message

```Python
 cond.setInitialComponentAmounts(b)

RuntimeError: ***ERROR***
Expecting a vector of initial amounts of conservative components with size 6 but given one has size 9 instead.
```

**Comment by allanleal** [comment]
Hi @taj004 , I could finally check the implementation of `KineticsSolver`. Currently, the specified component amounts in the `EquilibriumConditions` object is not being used in `KineticsSolver`. This needs to be corrected. I'll create an issue for this.

**Reply by taj004** [reply]
Hello @allanleal, Thank you for the help.

**Reply by allanleal** [reply]
I'm trying a temporary solution for you here. It would involve the use of `EquilibriumSolver` to equilibrate the aqueous species with given amounts of elements and `EquilibriumRestrictions` to prevent the minerals from reacting. After this, `KineticsSolver` is applied on the state whose aqueous species have been equilibrated.

**Reply by allanleal** [reply]
I've made the necessary changes in KineticsSolver to fix this behavior. #323 

In the next hour or so there should be a new conda package, v2.7.5 with these fixes. Running your script, I can see the amount of elements changing over time:

![image](https://github.com/reaktoro/reaktoro/assets/5825588/ffaf0ac5-86ee-47cf-805a-dd8ad09178ad)

**Reply by taj004** [reply]
Alright. Thank you for all your help.

**Comment by allanleal** [comment]
[Reaktoro v2.7.5](https://anaconda.org/conda-forge/reaktoro) resolves this issue. Please reopen the discussion otherwise. You can ignore the discussion above on the extent of reaction. You only need to provide the element amounts in the `b` array. However, I've identified that the best would be to also provide those values, but this will be corrected in a future major revision of the library and be properly explained.
